File: android/other/ProcessTest/app/src/main/java/com/example/processtest/MainActivity.java
package com.example.processtest;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.Handler;
import android.os.IBinder;
import android.os.Looper;
import android.os.Message;
import android.os.Messenger;
import android.os.RemoteException;
import android.util.Log;
import android.view.View;

public class MainActivity extends AppCompatActivity {

    public static int TEST_CODE = 1;
    private static final String TAG = "MainActivity";


    private final Handler handler = new Handler(Looper.getMainLooper()) {
        @Override
        public void handleMessage(@NonNull Message msg) {
            super.handleMessage(msg);
            if (msg.what == 11) {
                Bundle data = msg.getData();
                String replyTo = data.getString("replyTo");
                Log.d(TAG, replyTo);
            }
        }
    };
    private final Messenger mMessenger = new Messenger(handler);

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        bindService(new Intent(this, MyService.class), new ServiceConnection() {

            @Override
            public void onServiceConnected(ComponentName name, IBinder service) {
                Messenger messenger = new Messenger(service);
                Message message = Message.obtain();
                message.what = 1;
                Bundle bundle = new Bundle();
                bundle.putString("test", "client message");
                message.setData(bundle);
                message.replyTo = mMessenger;
                try {
                    messenger.send(message);
                } catch (RemoteException e) {
                    throw new RuntimeException(e);
                }
            }

            @Override
            public void onServiceDisconnected(ComponentName name) {

            }
        }, Context.BIND_AUTO_CREATE);
    }

    public void goSecondActivity(View view) {
        Intent intent = new Intent(this, SecondActivity.class);
        Bundle bundle = new Bundle();
        bundle.putString("test", "hello");
        intent.putExtras(bundle);
        startActivity(intent);
    }

    public void changeNum(View view) {
        TEST_CODE++;
        Log.d(TAG, String.valueOf(TEST_CODE));
    }
}

File: android/other/SaveInstanceStateTest/app/src/main/java/com/example/saveinstancestatetest/MainActivity.kt
package com.example.saveinstancestatetest

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log

private const val TAG = "MainActivity"

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        Log.d(TAG, "onCreate")
        savedInstanceState?.getString("tempData")?.let { Log.d("MainActivity", it) }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        val data = "设备发生横竖屏切换，或者点击 Home 键回到桌面时，会触发这个方法"
        outState.putString("tempData", data)
        Log.d(TAG, "onSaveInstanceState")
    }

    override fun onStop() {
        super.onStop()
        Log.d(TAG, "onStop")
    }

    override fun onDestroy() {
        super.onDestroy()
        Log.d(TAG, "onDestroy")
    }
}